### PR TITLE
Optional client certs

### LINKED ISSUE #18681: Doc optional client certificates
Add a section about optional client certificates, e.g. only requiring them for some portions of the application. See https://github.com/dotnet/aspnetcore/issues/21193#issuecomment-622456661
There are some related improvements to the cert auth handler in 5.0 where we added an OnChallenge event.
https://github.com/dotnet/aspnetcore/pull/21484

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 35b7960d-6de5-78ee-bc45-d7218239bbc8
* Version Independent ID: d2db534b-198a-dc99-5b5a-0525582d39ac
* Content: [Configure certificate authentication in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1)
* Content Source: [aspnetcore/security/authentication/certauth.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/authentication/certauth.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @blowdart
* Microsoft Alias: **bdorrans**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #18681

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.0&branch=pr-en-us-18684#optional-client-certificates)

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@Tratcher  is this in the right direction or do you want to provide a draft?

File: aspnetcore/security/authentication/certauth.md [hunk 1]
```diff
@@ -548,3 +548,19 @@ namespace AspNetCoreCertificateAuthApi
...
 ```
+
+<a name="occ"></a>
+
+## Optional client certificates
+
+This section provides information for apps that must protect a subset of the app with a certificate. For example, a Razor Page or controller in the app might require client certificates. This presents challenges as client certificates:
+  
+* Are a TLS feature, not an HTTP feature.
+* Are negotiated per-connection and must be be negotiated at the start of the connection before any HTTP data is available. At the start of the connection, only the Server Name Indication (SNI)&dagger; is known. The first request on a connection negotiates the certificate and further requests generally won't be able to renegotiate. Renegotiation is prohibited in HTTP/2.
```
Comment (Tratcher) [inline aspnetcore/security/authentication/certauth.md:559]:
```suggestion
* Are negotiated per-connection and must be be negotiated at the start of the connection before any HTTP data is available. At the start of the connection, only the Server Name Indication (SNI)&dagger; is known. The client and server certificates are negotiated prior to the first request on a connection and requests generally won't be able to renegotiate. Renegotiation is prohibited in HTTP/2.
```

**Comment by Rick-Anderson** [comment]
Should I add/edit the following:

You set up bindings for my.domain and clientcert.my.domain on the same site. The base site doesn't require a client cert but the subdomain does.
If you get to a part of the site that requires a client cert and doesn't have one, you redirect to the same page using the clientcert.my.domain host.
Since it's a different host name the client will establish a different connection and the client cert will be provided.
To sign out you redirect back to the my.domain version of the site with a connection: close header to close the old connection behind you.

**Comment by Tratcher** [comment]
Do you think people will be able to follow those steps without some concrete examples? Or should we include a sample using Kestrel or IIS? Or at least link to other docs that cover how you'd accomplish each step?
1. link to server docs
1. Link to authorization. Give an example of the 5.0 improvements to the cert auth handler for Challenge.
1. (just an observation)
1. Drop the part about sign-out. I realized it's misleading because it doesn't actually remove any state.

**Comment by Rick-Anderson** [comment]
@Tratcher  
> Do you think people will be able to follow those steps 

I added a link to a discussion alias to find out what we should add.

> Or at least link to other docs that cover how you'd accomplish each step?

Please provide links.

File: aspnetcore/security/authentication/certauth.md [hunk 2]
```diff
@@ -548,3 +548,28 @@ namespace AspNetCoreCertificateAuthApi
...
+* Are a TLS feature, not an HTTP feature.
+* Are negotiated per-connection and must be be negotiated at the start of the connection before any HTTP data is available. At the start of the connection, only the Server Name Indication (SNI)&dagger; is known. The client and server certificates are negotiated prior to the first request on a connection and requests generally won't be able to renegotiate. Renegotiation is prohibited in HTTP/2.
+
+ASP.NET Core 5 preview 4 and later supports optional client certificates. For more information, see the [Optional certificates sample](https://github.com/dotnet/aspnetcore/tree/master/src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample).
+
+The following approach supports optional client certificates:
+
+* Set up binding for the domain and subdomain:
+  * For example, set up bindings on `contoso.com` and `myClient.contoso.com`. The `contoso.com` host doesn't require a client certificate but `myClient.contoso.com` does.
+  * For more information, see `[TODO - provide links]()`
```
Comment (Tratcher) [inline aspnetcore/security/authentication/certauth.md:567]: See for Kestrel https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#listenoptionsusehttps and HttpsConnectionAdapterOptions.ClientCertificateMode 

And for IIS https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/?view=aspnetcore-3.1#create-the-iis-site and https://docs.microsoft.com/en-us/iis/manage/configuring-security/how-to-set-up-ssl-on-iis#configure-ssl-settings-2

For Http.Sys see https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/httpsys?view=aspnetcore-3.1#configure-windows-server and https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2008-R2-and-2008/cc725882(v=ws.10)?redirectedfrom=MSDN#add-sslcert clientcertnegotiation

File: aspnetcore/security/authentication/certauth.md [hunk 3]
```diff
@@ -548,3 +548,28 @@ namespace AspNetCoreCertificateAuthApi
...
+<a name="occ"></a>
+
+## Optional client certificates
+
+This section provides information for apps that must protect a subset of the app with a certificate. For example, a Razor Page or controller in the app might require client certificates. This presents challenges as client certificates:
+  
+* Are a TLS feature, not an HTTP feature.
+* Are negotiated per-connection and must be be negotiated at the start of the connection before any HTTP data is available. At the start of the connection, only the Server Name Indication (SNI)&dagger; is known. The client and server certificates are negotiated prior to the first request on a connection and requests generally won't be able to renegotiate. Renegotiation is prohibited in HTTP/2.
+
+ASP.NET Core 5 preview 4 and later supports optional client certificates. For more information, see the [Optional certificates sample](https://github.com/dotnet/aspnetcore/tree/master/src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample).
```
[the remark this answers is not in the document]
Comment (Rick-Anderson) [reply aspnetcore/security/authentication/certauth.md]: > The docs don't link to aspnetcore repo samples anywhere else do they?
See [Authentication samples for ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/samples?view=aspnetcore-3.1) 
In many place we point to source code.

https://raw.githubusercontent.com/dotnet/AspNetCore.Docs/master/aspnetcore/security/authentication/samples.md shows released links. I sometime use a permalink if I need to highlight a line.

I know master will change,  but hopefully for the better. I was planning to use master temporarily until we have a stand alone doc sample.  
I'll change to permalink now.

File: aspnetcore/security/authentication/certauth.md [hunk 4]
```diff
@@ -548,3 +548,35 @@ namespace AspNetCoreCertificateAuthApi
...
+ASP.NET Core 5 preview 4 and later adds more convenient support for optional client certificates. For more information, see the [Optional certificates sample](https://github.com/dotnet/aspnetcore/tree/9ce4a970a21bace3fb262da9591ed52359309592/src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample).
+
+The following approach supports optional client certificates:
+
+* Set up binding for the domain and subdomain:
+  * For example, set up bindings on `contoso.com` and `myClient.contoso.com`. The `contoso.com` host doesn't require a client certificate but `myClient.contoso.com` does.
+  * For more information, see:
+    * [Kestrel](/fundamentals/servers/kestrel):
+      * [ListenOptions.UseHttps](xref:fundamentals/servers/kestrel#listenoptionsusehttps)
+      * <xref:Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions.ClientCertificateMode>
```
Comment (Tratcher) [inline aspnetcore/security/authentication/certauth.md:570]:
```suggestion
      * <xref:Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions.ClientCertificateMode>
      * Note Kestrel does not currently support multiple TLS configurations on one binding, you'll need two bindings with unique IPs or ports. See https://github.com/dotnet/runtime/issues/31097
```

**Pull request merged**